src/Npgsql/NpgsqlDataSource.cs

```diff
@@ -397,34 +397,46 @@ internal virtual bool TryRemovePendingEnlistedConnector(NpgsqlConnector connecto
     #region Dispose
 
     /// <inheritdoc />
-    protected override void Dispose(bool disposing)
+    protected sealed override void Dispose(bool disposing)
     {
         if (disposing && Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
+            DisposeBase();
+    }
+
+    /// <inheritdoc cref="Dispose" />
+    protected virtual void DisposeBase()
+    {
+        var cancellationTokenSource = _timerPasswordProviderCancellationTokenSource;
+        if (cancellationTokenSource is not null)
         {
-            var cancellationTokenSource = _timerPasswordProviderCancellationTokenSource;
-            if (cancellationTokenSource is not null)
-            {
-                cancellationTokenSource.Cancel();
-                cancellationTokenSource.Dispose();
-            }
+            cancellationTokenSource.Cancel();
+            cancellationTokenSource.Dispose();
+        }
 
-            _passwordProviderTimer?.Dispose();
+        _passwordProviderTimer?.Dispose();
 
-            _setupMappingsSemaphore.Dispose();
+        _setupMappingsSemaphore.Dispose();
 
-            Clear();
-        }
+        Clear();
     }
 
     /// <inheritdoc />
-    protected override ValueTask DisposeAsyncCore()
+    protected sealed override ValueTask DisposeAsyncCore()
     {
         // TODO: async Clear, #4499
-        Dispose(true);
+        if (Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
+            DisposeAsyncBase();
 
         return default;
     }
 
+    /// <inheritdoc cref="DisposeAsyncCore" />
+    protected virtual ValueTask DisposeAsyncBase()
```

Comment:
This is starting to be lots of dispose methods :sweat_smile:

Maybe it's OK to just duplicate the DisposeBase code in DisposeAsyncCore and Dispose (especially since we need to call DisposeAsync on the password provider timer)

Reply:
Well, the main issue is, both `Dispose` and `DisposeAsyncCore` call `Interlocked.CompareExchange` to make sure there is no concurrent dispose in process. The point here was that `NpgsqlDataSource` makes sure to call dispose overrides only once.

Reply:
OK with me